### PR TITLE
fix(flow): infer table schema correctly

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?
- fix infer table schema for flow

Please explain IN DETAIL what the changes are in this PR and why they are needed:

previously, flow can't correctly infer pk&time index for query with `GROUP BY`& `tumble`, now function to infer those &related tests are added

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/flow/src/adapter.rs [hunk 1]
```diff
@@ -345,16 +343,32 @@ impl FlownodeManager {
...
+                        let ret = ColumnSchema::new(name, typ.scalar_type, typ.nullable);
+                        if schema.typ().time_index == Some(idx) {
+                            ret.with_time_index(true)
+                        } else {
+                            ret
+                        }
                     })
                     .collect_vec();
 
                 let mut with_ts = wout_ts.clone();
```
Comment (zhongzc) [inline src/flow/src/adapter.rs:357]: the name `wout_ts` will not be suitable.

File: src/flow/src/adapter.rs [hunk 2]
```diff
@@ -345,16 +343,32 @@ impl FlownodeManager {
...
+                    let ts_col = ColumnSchema::new(
+                        AUTO_CREATED_PLACEHOLDER_TS_COL,
+                        ConcreteDataType::timestamp_millisecond_datatype(),
+                        true,
+                    )
+                    .with_time_index(true);
+                    with_ts.push(ts_col);
+                }
+
+                (primary_keys, with_ts, schema.typ().time_index.is_none())
```
Comment (zhongzc) [inline src/flow/src/adapter.rs:371]: Here `schema.typ().time_index.is_none()` is always true, which cannot indicate that the ts column is a plaeholder

File: src/flow/src/adapter.rs [hunk 3]
```diff
@@ -508,12 +522,13 @@ impl FlownodeManager {
         debug!("Starting to run");
         loop {
             // TODO(discord9): only run when new inputs arrive or scheduled to
-            debug!("call run_available in run every second");
-            self.run_available(true).await.unwrap();
-            debug!("call send_writeback_requests in run every second");
+            if let Err(err) = self.run_available(true).await {
+                common_telemetry::error!("Run available errors: {:?}", err);
```
Comment (zhongzc) [inline src/flow/src/adapter.rs:526]:
```suggestion
                common_telemetry::error!(err; "Run available error");
```

File: src/flow/src/adapter.rs [hunk 4]
```diff
@@ -508,12 +522,13 @@ impl FlownodeManager {
...
+            }
             // TODO(discord9): error handling
-            self.send_writeback_requests().await.unwrap();
-            debug!("call log_all_errors in run every second");
+            if let Err(err) = self.send_writeback_requests().await {
+                common_telemetry::error!("Send writeback request errors: {:?}", err);
```
Comment (zhongzc) [inline src/flow/src/adapter.rs:530]:
```suggestion
                common_telemetry::error!(err; "Send writeback request error");
```

File: src/flow/src/repr/relation.rs [hunk 5]
```diff
@@ -132,13 +133,24 @@ impl RelationType {
...
+                        None
+                    } {
+                        old = *prev;
+                        if old < mfp.input_arity {
+                            break;
+                        }
+                    }
+                    (old, new)
+                },
+            ));
```
Comment (zhongzc) [inline src/flow/src/repr/relation.rs:153]: Can you extract it as a function and provide some comments to explain why?

File: src/flow/src/adapter.rs (hunk 2 above)
Comment (discord9) [reply src/flow/src/adapter.rs]: > Here `schema.typ().time_index.is_none()` is always true, which cannot indicate that the ts column is a plaeholder

Added a `find_time_index_in_group_exprs` function to assign time index now, so there is cases where `time_index.is_some()`

File: src/flow/src/repr/relation.rs (hunk 5 above)
Comment (discord9) [reply src/flow/src/repr/relation.rs]: > Can you extract it as a function and provide some comments to explain why?

Extracted and added more comments

**Review by zhongzc — APPROVED** [review]

File: src/flow/src/transform/plan.rs [hunk 6]
```diff
@@ -302,9 +203,110 @@ impl TypedPlan {
...
+            } else {
+                None
+            }
+        })
+        .collect();
+    for (key_idx, key_expr) in key_exprs.iter().enumerate() {
+        if let ScalarExpr::Column(_) = key_expr {
+            if !all_cols_ref_in_proj.contains(&key_idx) {
+                return InvalidQuerySnafu{
+                    reason: format!("Expect normal column in group by also appear in projection, but column {}(name is {}) is missing", key_idx, reduce_output_type.get_name(key_idx).clone().map(|s|format!("'{}'",s)).unwrap_or("unknown".to_string()))
```
Comment (waynexia) [inline src/flow/src/transform/plan.rs:297]: sugg: write this into two statements. `format!` macro cannot be formatted properly

**Review by waynexia — APPROVED** [review]
Mostly believe good to me 🫣

**Pull request merged**